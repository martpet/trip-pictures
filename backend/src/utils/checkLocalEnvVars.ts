import { EnvName } from '~/types';

export const checkLocalEnvVars = (envName: EnvName) => {
  const names = [
    'VITE_PERSONAL_SUBDOMAIN',
    'PERSONAL_DEV_ACCOUNT_ID',
    'PERSONAL_DEV_ACCOUNT_REGION',
  ];
  if (envName === 'personal') {
    names.forEach((varName) => {
      if (!process.env[varName]) {
        throw new Error(`environment variable "${varName}" is missing from .env.local`);
      }
    });
  }
};
