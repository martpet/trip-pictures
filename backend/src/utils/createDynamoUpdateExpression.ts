export const createDynamoUpdateExpression = (
  data: Record<string, unknown>,
  parentKey?: string
) => {
  const ExpressionAttributeNames: Record<string, string> = {};
  const ExpressionAttributeValues: Record<string, unknown> = {};
  const actions: string[] = [];

  Object.entries(data).forEach(([attr, val]) => {
    actions.push(`${parentKey ? `#${parentKey}.` : ''}#${attr} = :${attr}`);
    ExpressionAttributeNames[`#${attr}`] = attr;
    ExpressionAttributeValues[`:${attr}`] = val;
    if (parentKey) {
      ExpressionAttributeNames[`#${parentKey}`] = parentKey;
    }
  });

  return {
    UpdateExpression: `set ${actions.join(',')}`,
    ExpressionAttributeNames,
    ExpressionAttributeValues,
  };
};
