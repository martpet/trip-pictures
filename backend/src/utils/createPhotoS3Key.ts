type Props = {
  userId: string;
  fingerprint: string;
};

export function createPhotoS3Key({ userId, fingerprint }: Props) {
  return `${userId}/${fingerprint}`;
}
