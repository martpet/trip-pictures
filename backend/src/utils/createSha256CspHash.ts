import crypto from 'crypto';

export const createSha256CspHash = (content: string) =>
  `sha256-${crypto.createHash('sha256').update(content).digest('base64')}`;
