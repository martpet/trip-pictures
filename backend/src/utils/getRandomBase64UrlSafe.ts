import crypto from 'crypto';
import { promisify } from 'util';

export const getRandomBase64UrlSafe = async (size: number) => {
  const cryptoRandomBytes = promisify(crypto.randomBytes);
  const bytes = await cryptoRandomBytes(size);
  return bytes.toString('base64url');
};
