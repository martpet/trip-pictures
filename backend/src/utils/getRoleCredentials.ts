import { AssumeRoleCommand, STSClient } from '@aws-sdk/client-sts';
import { Credentials } from '@aws-sdk/types';

export const getRoleCredentials = async (RoleArn: string, sessionName: string) => {
  const stsClient = new STSClient({});
  const camelcaseKeys = (await import('camelcase-keys')).default;

  const assumeRoleCommand = new AssumeRoleCommand({
    RoleArn,
    RoleSessionName: `${sessionName}-${new Date().getTime()}`,
  });

  const { Credentials: credentials } = await stsClient.send(assumeRoleCommand);

  if (!credentials) {
    throw Error('Cannot get credentials');
  }

  if (!credentials.AccessKeyId || !credentials.SecretAccessKey) {
    throw Error('Missing credentials prop');
  }

  return camelcaseKeys(credentials) as Credentials;
};
