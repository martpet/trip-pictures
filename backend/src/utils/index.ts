export * from '~/../../shared/utils';
//
export * from './checkLocalEnvVars';
export * from './createDynamoUpdateExpression';
export * from './createPhotoS3Key';
export * from './createSha256CspHash';
export * from './getAllowedOrigins';
export * from './getRandomBase64UrlSafe';
export * from './getRoleCredentials';
export * from './objectValuesToJson';
