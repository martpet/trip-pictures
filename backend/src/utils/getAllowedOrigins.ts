import { appEnvs, localhostUrl } from '~/consts';
import { getCdkEnv } from '~/utils';

export const getAllowedOrigins = () => {
  const envName = getCdkEnv();
  const { appDomain } = appEnvs[envName];

  const origins = [`https://${appDomain}`];

  if (envName === 'personal') {
    origins.push(localhostUrl);
  }

  return origins;
};
