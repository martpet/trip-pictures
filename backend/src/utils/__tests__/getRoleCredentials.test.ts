import { AssumeRoleCommand, Credentials, STSClient } from '@aws-sdk/client-sts';
import { mockClient } from 'aws-sdk-client-mock';
import camelcaseKeys from 'camelcase-keys';

import {
  itCalls,
  itRejects,
  itResolves,
  itSendsAwsCommand,
} from '~/constructs/Api/utils';
import { getRoleCredentials } from '~/utils';

const stsMock = mockClient(STSClient);

vi.mock('camelcase-keys');

const args = ['dummyRoleArn', 'dummySessionName'] as Parameters<
  typeof getRoleCredentials
>;

const assumeRoleOutput = {
  Credentials: {
    AccessKeyId: 'dummyAccessKeyId',
    SecretAccessKey: 'dummySecretAccessKey',
    SessionToken: 'dummySessionToken',
    Expiration: new Date(),
  },
};

beforeEach(() => {
  stsMock.reset();
  stsMock.on(AssumeRoleCommand).resolves(assumeRoleOutput);
});

describe('getRoleCredentials', () => {
  itCalls(camelcaseKeys, getRoleCredentials, args);
  itSendsAwsCommand(AssumeRoleCommand, stsMock, getRoleCredentials, args);
  itResolves(getRoleCredentials, args);

  describe('when credentials are missing from "AssumeRoleCommand" output', () => {
    beforeEach(() => {
      stsMock.on(AssumeRoleCommand).resolves({});
    });
    itRejects(getRoleCredentials, args);
  });

  describe.each(['AccessKeyId', 'SecretAccessKey'])(
    'when "%s" credentials prop is missing from "AssumeRoleCommand" output',
    (key) => {
      const assumeRoleOutputClone = structuredClone(assumeRoleOutput);
      delete assumeRoleOutputClone.Credentials[key as keyof Credentials];

      it('rejects with a correct value', async () => {
        await stsMock.on(AssumeRoleCommand).resolves(assumeRoleOutputClone);
        return expect(getRoleCredentials(...args)).rejects.toMatchSnapshot();
      });
    }
  );
});
