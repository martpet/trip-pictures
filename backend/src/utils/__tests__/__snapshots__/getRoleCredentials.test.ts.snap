// Vitest Snapshot v1

exports[`getRoleCredentials > calls "camelcase-keys" with correct args 1`] = `
[
  [
    {
      "AccessKeyId": "dummyAccessKeyId",
      "Expiration": 2022-01-01T00:00:00.000Z,
      "SecretAccessKey": "dummySecretAccessKey",
      "SessionToken": "dummySessionToken",
    },
  ],
]
`;

exports[`getRoleCredentials > resolves with a correct value 1`] = `
{
  "dummyCamelCaseKey": "dummyCamelCaseValue",
}
`;

exports[`getRoleCredentials > sends "AssumeRoleCommand" from "STSClient" with correct args 1`] = `
{
  "RoleArn": "dummyRoleArn",
  "RoleSessionName": "dummySessionName-1640995200000",
}
`;

exports[`getRoleCredentials > when "AccessKeyId" credentials prop is missing from "AssumeRoleCommand" output > rejects with a correct value 1`] = `[Error: Missing credentials prop]`;

exports[`getRoleCredentials > when "SecretAccessKey" credentials prop is missing from "AssumeRoleCommand" output > rejects with a correct value 1`] = `[Error: Missing credentials prop]`;

exports[`getRoleCredentials > when credentials are missing from "AssumeRoleCommand" output > rejects with a correct value 1`] = `[Error: Cannot get credentials]`;
